samples/bluetooth/fast_pair/locator_tag/src/dfu.c

```diff
@@ -0,0 +1,194 @@
+		return;
+	}
+
+	LOG_INF("DFU: mode %sabled", new_mode ? "en" : "dis");
+
+	dfu_mode = new_mode;
+
+	app_fp_adv_request(&fp_adv_trigger_dfu, dfu_mode);
+
+	/* Ensure that the advertising payload is updated if the advertising was already
+	 * enabled before the advertising request.
+	 */
+	if ((pre_request_fp_adv_mode == APP_FP_ADV_MODE_DISCOVERABLE) ||
+	    (pre_request_fp_adv_mode == APP_FP_ADV_MODE_NOT_DISCOVERABLE)) {
+		app_fp_adv_payload_refresh();
+	}
+
+	app_ui_state_change_indicate(APP_UI_STATE_DFU_MODE, dfu_mode);
+}
+
+static void dfu_factory_reset_prepare(void)
+{
+	dfu_mode_change(false);
+}
+
+APP_FACTORY_RESET_CALLBACKS_REGISTER(factory_reset_cbs, dfu_factory_reset_prepare, NULL);
+
+bool app_dfu_bt_gatt_operation_allow(const struct bt_uuid *uuid)
+{
+	if (bt_uuid_cmp(uuid, BT_UUID_SMP_CHAR) != 0) {
+		return true;
+	}
+
+	if (!dfu_mode) {
+		LOG_WRN("DFU: SMP characteristic access denied, DFU mode is not active");
+		return false;
+	}
+
+	(void) k_work_reschedule(&dfu_mode_timeout_work, K_MINUTES(DFU_MODE_TIMEOUT_MIN));
+
+	return true;
+}
+
+/* Due to using legacy advertising set size, add SMP UUID to either AD or SD,
+ * depending on the space availability related to the advertising mode.
+ * Otherwise, the advertising set size would be exceeded and the advertising
+ * would not start.
+ * The SMP UUID can be added only to one of the data sets.
+ */
+
+static int get_ad_data(struct bt_data *ad, const struct bt_le_adv_prov_adv_state *state,
+		    struct bt_le_adv_prov_feedback *fb)
+{
+	ARG_UNUSED(state);
+	ARG_UNUSED(fb);
+
+	enum app_fp_adv_mode current_fp_adv_mode = app_fp_adv_mode_get();
+
+	if (!dfu_mode) {
+		return -ENOENT;
+	}
+
+	if (current_fp_adv_mode != APP_FP_ADV_MODE_DISCOVERABLE) {
+		return -ENOENT;
+	}
+
+	*ad = data;
+
+	return 0;
+}
+
+static int get_sd_data(struct bt_data *sd, const struct bt_le_adv_prov_adv_state *state,
+		    struct bt_le_adv_prov_feedback *fb)
+{
+	ARG_UNUSED(state);
+	ARG_UNUSED(fb);
+
+	enum app_fp_adv_mode current_fp_adv_mode = app_fp_adv_mode_get();
+
+	if (!dfu_mode) {
+		return -ENOENT;
+	}
+
+	if (current_fp_adv_mode != APP_FP_ADV_MODE_NOT_DISCOVERABLE) {
+		return -ENOENT;
+	}
+
+	*sd = data;
+
+	return 0;
+}
+
+/* Used in discoverable adv */
+BT_LE_ADV_PROV_AD_PROVIDER_REGISTER(smp_ad, get_ad_data);
+
+/* Used in the not-discoverable adv */
+BT_LE_ADV_PROV_SD_PROVIDER_REGISTER(smp_sd, get_sd_data);
+
+static void dfu_mode_action_handle(void)
+{
+	if (dfu_mode) {
+		LOG_INF("DFU: refreshing the DFU mode timeout");
+	} else {
+		LOG_INF("DFU: entering the DFU mode for %d minute(s)",
+			DFU_MODE_TIMEOUT_MIN);
+	}
+
+	(void) k_work_reschedule(&dfu_mode_timeout_work, K_MINUTES(DFU_MODE_TIMEOUT_MIN));
+
+	dfu_mode_change(true);
+}
+
+static void dfu_mode_timeout_work_handle(struct k_work *w)
+{
+	LOG_INF("DFU: timeout expired");
+
+	dfu_mode_change(false);
+}
+
+void app_dfu_fw_version_log(void)
+{
+	LOG_INF("DFU: Firmware version: %s", CONFIG_MCUBOOT_IMGTOOL_SIGN_VERSION);
```

Comment:
Consider using `APP_VERSION_TWEAK_STRING` from `app_version.h` header (to consistently rely on the header content for versioning information)